runtime/src/bank.rs

```diff
@@ -8115,30 +8135,81 @@ impl Bank {
+                // account, then we want to update the existing data account
+                if let Some(new_data_account) = self.get_account_with_fixed_root(&new_data_address)
+                {
+                    // A data account exists for the new program
+                    // Check if the old program account has a data account
+                    if let Some(old_data_account) =
+                        self.get_account_with_fixed_root(&old_data_address)
+                    {
+                        // It does. Replace it with the new data account
+                        self.replace_account(
+                            &old_data_address,
+                            &new_data_address,
+                            Some(&old_data_account),
+                            &new_data_account,
+                        );
+                        // The old program account will already house the PDA
+                        // of the data account
+                    } else {
+                        // It does _not_. Create it with the new data account
+                        self.replace_account(
+                            &old_data_address,
+                            &new_data_address,
+                            None,
+                            &new_data_account,
+                        );
+                        // Update the old program account to house the PDA of
+                        // the data account
+                        {
+                            let mut account = Account::from(old_account.clone());
+                            account.data = old_data_address.as_ref().to_vec();
+                            let account_shared_data = AccountSharedData::from(account);
+                            self.store_account(old_address, &account_shared_data);
+                        }
+                    }
+                    // We only swapped the data accounts, so now we need to
+                    // clear the new program account
+                    self.capitalization
+                        .fetch_sub(new_account.lamports(), Relaxed);
+                    self.store_account(new_address, &AccountSharedData::default());
+                } else if self
+                    .get_account_with_fixed_root(&old_data_address)
+                    .is_none()
+                {
+                    // A data account does not exist for the new program
+                    // Swap program accounts only
+                    self.replace_account(
+                        old_address,
+                        new_address,
+                        Some(&old_account),
+                        &new_account,
+                    );
+                }
 
                 // Unload a program from the bank's cache
                 self.loaded_programs_cache
                     .write()
                     .unwrap()
                     .remove_programs([*old_address].into_iter());
```

Comment:
We should also test to make sure this doesn't have any issues if the old program didn't exist before. I'm imagining we just need to skip it and we'll be good